### ISSUE TITLE
ZoL appears to write to reserved bootblock area and thus, corrupting it

### ISSUE THREAD

**Issue by FransUrbo** [zfsonlinux/zfs#3207]
I've been getting a lot of help from tsoome (on IRC) to get a new Grub working with the new ZoL features. I don't fully understand the issue, but I'm trying to include as much as possible. Much of this is his wording, which I have tried to understand. Without fully succeeding.

This is mostly working (?) but there seems to be a problem with grub embedding.

What I did:

1. Booting the (unofficial) Debian GNU/Linux Wheezy installer image on a VBox VM.
    * This image have native support for ZFS
2. Create a GPT partition table on each disk, giving ZFS the full-disk device node.
3. Create the filesystem hierarchy - <code>rpool/ROOT/debian{/,/boot,/home,/usr,/var}</code>
4. Install the system
5. Write the grub header (?) to <code>/dev/sda</code>
6. Reboot
    * This works just fine, the system comes up as it should
7. Reboot
    * The VM gets a "Guru Meditation" from VBox!

To make sure that it wasn't the OS (somewhere, somehow), I did it all from scratch, but started the system (in point 6) with <code>init=/bin/bash</code> effectively making SURE that nothing messes with it (not that I thought it did, but it was a possibility) and then executed point 7.

And after this, the machine would never again boot (always resulting in "Guru Meditation"). BUT, booting from the install image, but instead of doing the partitioning etc, I simply switch to a free console, do a scrub on the (already) imported pool, mount all filesystems again (as in point one below) and once the scrub is complete, run <code>grub-install /dev/sda</code> again. It would then survive ONE reboot, but again crash/hang/guru on the next...

To verify that this is actually a problem, I did the following:

1. Import pool

    ```
    zpool import -f -o altroot=/target -a
    zfs mount -a
    for fs in boot home usr var; do mount -t zfs rpool/ROOT/debian/$fs /target/$fs; done
    for fs in sys dev proc; do mount -o bind /$fs /target/$fs; done
    ```

2. Read bootblock

    ```
    dd if=/dev/sda  of=sda-1.dd bs=512 skip=1024 count=6144
    ```

3. Write grub

    ```
    chroot /target /bin/bash
    grub-install /dev/sda
    ```

4. Read bootblock

    ```
    dd if=/dev/sda  of=sda-2.dd bs=512 skip=1024 count=6144
    ```

5. Make changes

    ```
    echo whatever > /target/tmp/x
    ```

6. Export pool

    ```
    zpool export rpool
    ```

7. Import pool

    ```
    zpool import -f -o altroot=/target -a
    ```

8. Read bootblock

    ```
    dd if=/dev/sda  of=sda-3.dd bs=512 skip=1024 count=6144
    ```

8. Compare files

    ```
    md5sum sda-1.dd sda-2.dd
    od -c sda-1.dd > sda-1.od
    od -c sda-2.dd > sda-2.od
    od -c sda-3.dd > sda-3.od
    diff -u sda-1.od sda-2.od
    diff -u sda-2.od sda-3.od
    ```

Both diffs gives _a lot_ of changes!

With a slightly modified <code>grub-probe</code> (which changed <code>s/grub_dprintf/grub_printf/g</code> in <code>zio_checksum_verify()</code>, I get this:

```
zfs: checksum label verification failed
zfs: actual checksum 36a731d2a1e575c9 e1e968c61a3dbc10 44ea72b46d993161 d5cb866922e78711
zfs: expected checksum 2cf4521d8951c45b 70236654e6f66cd9 7fb14e25e0127100 3e3fc38c4a0a56e4
zfs: checksum label verification failed
zfs: actual checksum 49eb917c86185b5c d1d1c664cc30ea66 8a7a63b5cbddeab6 9515d34796708760
zfs: expected checksum a217ec51f43f8309 e5754a35dab96b9b 14e5206e6185a6ee 56af5fa83285b84f
zfs: checksum label verification failed
zfs: actual checksum 0a6a690ad284a66b 78ae6c1aeda2c2ef 64f59e8536b3cacc f0d6e5c182bf6882
zfs: expected checksum 2f180758f4e5528a fe1ffec61d6aa424 9bc2eab6971ff39a d101914b7ea7f0c5
zfs: checksum label verification failed
zfs: actual checksum e5aeacc21091ddf9 496e838230c586c3 d32a6639e7cdb88a ea0ca86f84df05a9
zfs: expected checksum 683519ca43c1b5ca 7daf537dd2fcb46b 023f8c56d11f0c79 ff1f11177990f2f0
zfs
```

Additional info:

```
[20:45:45] <tsoome> for explanation: this area is reserved to be used by bootloader, to store boot program, its size is 3.5MB and area starts right after second label (page 14, ondiskformat0822.pdf)
[20:49:49] <tsoome> also, the defect should actually appear on all pools, its nothing special for your boot pool, but it's unclear, why this happens
[20:50:50] <tsoome> if ashift is playing tricks, if actual write in vdev_label_write will be larger or some other reason, it just has to be found out
[20:54:54] <tsoome> and just to emphasise; even illumos only recently (few weeks ago) started to use this area, so far I haven't seen reports about problems with it, so its quite possible its zol specific issue. or, maybe even vbox+linux+zol specific issue, altho i don't believe vbox is playing any role here.
```

Apparently, according to tsoome, the offending code is in https://github.com/zfsonlinux/zfs/blob/master/module/zfs/vdev_label.c#L1014-1018

**Comment by aarcane** [comment]
This is an unsupported setup. You need to be sure to add a bios-grub partition to each disk in your rpool. Read any zfs-root guide and they'll tell you about this, unless they're using legacy mbr partition tables, in which case it doesn't matter.

**Comment by FransUrbo** [comment]
It's only unsupported because no one have written this yet. But tsoome did and I put it in a grub package for ZoL.

So it is now supported by grub (both ZoL and Illumos grub) and ZFS in Illumos, but not by ZoL. Hence the issue report.

**Comment by tsoome** [comment]
its unsupported setup for grub which does not implement embedding for zfs; still the bootblock area is reserved space for zfs and therefore off limits, zfs must not write anything there. if it still does, its bug. also note that the issue was only found because the updated grub was failing there. zfs on disk format, page 14.

**Comment by ryao** [comment]
From IRC:

```
22:28 < FransUrbo> ryao: Have time to take a quick lock at #3207
22:28 < zfs>     ZoL appears to write to reserved bootblock area and thus, corrupting it . Issue #3207 . zfsonlinux/zfs . GitHub https://github.com/zfsonlinux/zfs/issues/3207
22:34 <+ryao> FransUrbo: This is the problem: https://github.com/zfsonlinux/zfs/blob/6e9710f7c33a7440904b39fb4e6b68b3325cfeca/include/sys/vdev_impl.h#L242
22:35 <+ryao> FransUrbo: It is 10:30pm here so this is not the best time for me to discuss this, but it looks to me like the back labels aren't being updated properly. It is writing them in sequence as if labels 2 and 3 are written sequentially on the disk with labels 0 
              and 1.
```

**Comment by behlendorf** [comment]
For future reference I wasn't able to reproduce this issue in a VM.  I created a pool on a new device which had the first 4M of filled with the pattern 0xffffffff.  Then I ran the following tiny script to force TXGs to be written so the uber blocks and labels would both be updated.

```
 for (( i=0; i<1000; i++ )); do sudo ./cmd/zpool/zpool import tank; sudo ./cmd/zfs/zfs mount -a; touch /tank/fish/file; rm /tank/fish/file; sudo ./cmd/zpool/zpool export tank; echo "pass $i"; done
```

Periodically I'd check the boot block by running `sudo dd if=/dev/vdb1 bs=1M count=4 | hexdump -C` to verify it was intact.  In all cases I saw the boot block still contained the original pattern.

```
...
00080000  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00400000
```

However, it certainly sounds like there is an issue here and I just wasn't able to reproduce it so more investigation is needed.
